interfaces/apparmor/apparmor.go

```diff
@@ -0,0 +1,94 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package apparmor contains primitives for working with apparmor.
+//
+// References:
+//  - http://wiki.apparmor.net/index.php/Kernel_interfaces
+//  - http://apparmor.wiki.kernel.org/
+//  - http://manpages.ubuntu.com/manpages/xenial/en/man7/apparmor.7.html
+package apparmor
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+// LoadProfile loads an apparmor profile from the given file.
+//
+// If no such profile was previously loaded then it is simply added to the kernel.
+// If there was a profile with the same name before, that profile is replaced.
+func LoadProfile(fname string) error {
+	return exec.Command("apparmor_parser", "--replace", fname).Run()
+}
+
+// Profile contains the name and mode of an apparmor profile loaded into the kernel.
+type Profile struct {
+	// Name of the profile. This is is either full path of the executable or an
+	// arbitrary string without spaces.
+	Name string
+	// Mode is either "enforcing" or "complaining".
+	Mode string
+}
+
+// Unload removes a profile from the running kernel.
+//
+// The operation is done with: apparmor_parser --remove $name
+func (profile *Profile) Unload() error {
+	return exec.Command("apparmor_parser", "--remove", profile.Name).Run()
+}
+
+// profilesPath contains information about the currently loaded apparmor profiles.
+const realProfilesPath = "/sys/kernel/security/apparmor/profiles"
+
+var profilesPath = realProfilesPath
+
+// LoadedProfiles interrogates the kernel and returns a list of loaded apparmor profiles.
+//
+// Snappy manages apparmor profiles named *.snap. Other profiles might exist on
```

Comment:
On Thu, 2016-03-10 at 12:27 -0800, Zygmunt Krynicki wrote:
> > 
> > +
> > +// Unload removes a profile from the running kernel.
> > +//
> > +// The operation is done with: apparmor_parser --remove $name
> > +func (profile *Profile) Unload() error {
> > +	return exec.Command("apparmor_parser", "--remove",
> > profile.Name).Run()
> > +}
> > +
> > +// profilesPath contains information about the currently loaded
> > apparmor profiles.
> > +const realProfilesPath = "/sys/kernel/security/apparmor/profiles"
> > +
> > +var profilesPath = realProfilesPath
> > +
> > +// LoadedProfiles interrogates the kernel and returns a list of
> > loaded apparmor profiles.
> > +//
> > +// Snappy manages apparmor profiles named *.snap. Other profiles
> > might exist on
> This is done so that we have a clear way to identify profiles managed
> by snappy. We discussed this with Gustavo last week / this week as
> well.
> 
I'm not sure this is the best way to do this. The kernel only has what
is loaded. The authoritative list of what snappy maintains is what is
in /var/lib/snappy/apparmor/profiles.

For example, if you remove a profile from the kernel, snappy is still
managing the profile even though it wouldn't show up
in /sys/kernel/security/apparmor/profiles. Plus, as is obvious from
what .snap is trying to handle, there are many other profiles you have
to filter out and this is buggy because a non-snappy managed profile
may end in .snap, especially on a system where the user has traditional
admin control, such as snappy dimension on Ubuntu classic. Ie, what
will happen if a user creates the profile 'oh.snap' on her system for
some home-grown software?

> I'm aware that nothing currently creates *.snap profiles but it is my
> desire to switch over ``interfaces/security.go`` to do so tomorrow.
> 
> I'd like to understand required changes better, I know that generated
> wrappers will have to change (so that we call it with
> ``$snap[.$app].snap`` as the policy name) but I'm not sure what else
> needs changing.
> 

1. The launcher is expecting a profile name. Adding '.snap' to the
profile name means we can either change the launcher to append it or we
can change the /snaps/bin generation and the systemd unit generation. I
prefer the latter.

2. profileAttachForApp() needs to change to append .snap

3. varsForApp() needs to change so that encoded '.snap' (_2esnap)
appears in APP_ID_DBUS

I prefer you abstract this out into a 'profileNameForApp()' (or
similar) that the above 3 points can use rather than sprinkling .snap
in various places.

I audited all existing snappy and Touch policy. In general, the snap
policy today should be ok because our policy doesn't care about the
version part of the profile name (which .snap then becomes part of) or
we use the apparmor var @{profile_name}. Touch has some rules that
would be affected by this change but not many. Touch trusted helpers I
think typically don't care about the version part of the profile name,
so they should be ok.

Moving forward, we might have to consider '.snap' as part of the label
for new policy when writing rules for new interfaces (eg, that use
dbus, unix, signal and ptrace and in the future, network).

That should be it but I maintain that adding .snap adds complexity
where I think it shouldn't necessarily be and the implementation is
prone to bugs.

-- 
Jamie Strandboge             | http://www.canonical.com